srml/session/src/lib.rs

```diff
@@ -51,19 +51,40 @@ pub trait OnSessionChange<T> {
 	fn on_session_change(time_elapsed: T, should_reward: bool);
 }
 
-impl<T> OnSessionChange<T> for () {
-	fn on_session_change(_: T, _: bool) {}
+macro_rules! for_each_tuple_ {
+    ($m:ident !!) => {
+        $m! { }
+    };
+    ($m:ident !! $h:ident, $($t:ident,)*) => {
+        $m! { $h $($t)* }
+        for_each_tuple_! { $m !! $($t,)* }
+    }
+}
+macro_rules! for_each_tuple {
+    ($m:ident) => {
+        for_each_tuple_! { $m !! A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, }
```

Comment:
I would propose to merge `for_each_tuple` and `for_each_tuple_`.
You could use a marker for the inner stuff, so something like:
```
macro_rules! for_each_tuple {
    ($m:ident) => {
        for_each_tuple! { @IMPL $m !! A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, }
    },
    (@IMPL $m:ident !!) => { $m! { } };
    (@IMPL $m:ident !! $h:ident, $($t:ident,)*) => {
        $m! { $h $($t)* }
        for_each_tuple! { @IMPL $m !! $($t,)* }
    }
}
```